### DISCUSSION TITLE
Intermediate submission feedback, timeline clarification and more

### DISCUSSION THREAD

**Discussion by fpetric** [larics/icuas24_competition#39] [Announcements]
Dear teams,

thank you to all who uploaded their solution to the intermediate deadline, and thus completed their registration! The teams that are in are (in no particular order):

## Final list of teams
- AIRo Lab
- ITUAV
- AVANT-UFMG
- AVADER AGH
- UMIC
- SDCNLab
- UniboUAV
- AeroMIT
- UoP
- Skoltech Hawks
- Aerial Robotics IITK
- high_figh_fly
- ASL
- Aerial Robotics Kharagpur
- FeikAir
- UAS-DTU
- CVAR
- TeamHawkeye
- LCIS
- Sandrone
- BDP Fly
- CapyBots
- percepDrone
- Democritus Mechatronics & Automation - DMA

## Some remarks for the future submissions

- Make sure to create additional pane in session.yml for your nodes (it is easier for us to check if your code is running at all)
- When adding ros commands in the tmux pane, use waitForRos/waitForSimulation etc. macros to wait for roscore or gazebo to appear, otherwise your nodes may not run
- Only launch your nodes by including them in session.yml file (do not launch them from start.sh script or similar), we will use different start script and if it is not in session.yml, it will not be started
- Make sure to printout something like `Node started, running or similar` upon launch of your code (the same reason as above, it is easier for us to check if it's working)
- Make sure your Python nodes have shebang line and are executable
- Make sure any other scripts you are expecting us to run are also executable (including docker_build and docker_run scripts)
- Do not change docker image and container name if possible
- Fruit count message should be published only once 
- If you are not using trajectories, be careful with position references that are far away from the current position of the UAV, the UAV controllers are aggressive by default and large position error = large speed = large pitch/roll angle = large loss of height = crash with the floor (crashes will be heavily penalized)

## What's next

- Sometime today, we will delete the files from your folders and they will be ready for your real solutions of the simulation phase. The procedure is as follows:
- You upload your solution at any time between tomorrow and the deadline announced in the email: **February 8th, 11:59 AOE**
- After the deadline, we download what is up there (you do not need to submit if you are not ready)
- We run your solution once. Run procedure will be as follows:
  - Download zip, extract, run docker_build.sh, run docker_run.sh, run start.sh inside the container
  - Run evaluation node in parallel, that measures time, tracks the UAV position/pose and waits for the fruit count to be published
  - Calculate points (exact measure TBD soon)
- You get feedback, we post results online
- You improve your code, or not, or leave the existing submission without update, and we reach the next evaluation deadline (February 15th unless we skip this one if the evaluation of the first one take too long), download the latest versions of your submissions and run all of them once with the same procedure as before, you get feedback, we post results online and so on.
- Next, the final simulation upload deadline is February 23rd, if you do not upload anything by this date, you get zero points. After this deadline we will run your final solution N (N is likely to be 3 or 5) times, and your overall result will be the average of best N runs. If you submit to the first evaluation window, the total number of runs for your solution will be N+2 (i.e. you can have a better run in the earlier evaluation window, and your worst two runs of all the runs will be discarded from calculation). This means that a bad run in the early eval window will not drop your score. 
 
## Post your questions here pls :)

**Comment by allenthreee** [comment]
"If you submit to the first evaluation window, the total number of runs for your solution will be N+2"
Does this mean NOT updating the origin solution submitted before 8th Feb will give the team 2 more runs?

**Reply by fpetric** [reply]
If you submit before the first deadline you get:

- 1 run in the first evaluation period
- 1 run in the second evaluation period (whether you update the solution or not)
- N runs after the final deadline
- Only N of those will count (2 will be discarded)

If you submit after the first and before the second deadline you get:

- 0 runs in the first evaluation period
- 1 run in the second evaluation period
- N runs after the final deadline
- Only N of those will count (1 will be discarded)

If you submit only at the final deadline, you get N runs, and the average of those is then calculated.

**Reply by tanzhengo** [reply]
Is the first evaluation period is any day from JAN 8th to JAN 15th ，the second evaluation is any day from JAN 15th to JAN 23rd?

**Reply by fpetric** [reply]
you need to submit by Feb 8th 11:59 AOE to be included in the first evaluation. Whether you submit now or on Feb 8th 11:55 AOE is the same.

**Reply by allenthreee** [reply]
So updating our code is allowed, right?

**Reply by fpetric** [reply]
Yes.

**Comment by allenthreee** [comment]
We run your solution once. Run procedure will be as follows:
 * Download zip, extract, run build_docker.sh, run run_docker.sh, run start.sh inside the container
 * Run evaluation node in parallel, that measures time, tracks the UAV position/pose and waits for the fruit count to be published
 * Calculate points (exact measure TBD soon)


The origin file name is "docker_build.sh", should we change it to "build_docker.sh" or is it a typo, and we still use "docker_build.sh"?

**Reply by fpetric** [reply]
Updated the instructions, leave filenames as is. Sorry.

**Reply by allenthreee** [reply]
Thanks a lot for the update:)
have you deleted the files in our onedrive folder?
And should we also name our new submission TEAMNAME_docker.zip?

**Reply by fpetric** [reply]
The folders should be clean now. `<teamname>_docker.zip` should be the name of the file, yes

**Reply by allenthreee** [reply]
Thank you for your swift reply.

One more question: Should the content be similar to intermediate submission, including session.yml . For example, the following image shows our content of the zip file. All our solution is in the folder AIRo_solution/ folder. Is it okay? 
![submission_content](https://github.com/larics/icuas24_competition/assets/59171742/55996e06-dcb8-40b5-9533-906aa795a4ce)

**Reply by fpetric** [reply]
As long as your docker setup (dockerfile and build/run scripts) prepares a valid workspace, and `start.sh` launches correct nodes through `session.yml`, you can organize your files however you want. As I mentioned sometime earlier, we will launch `start.sh` from the `icuas24_competition` package inside the container. Where are the other files, nodes, packages and so on is not important, as long as the container is set up correctly through docker build and run scripts.

**Reply by AR-AK** [reply]
Thank you for the update, May you please elaborate, Does your statement "We will Run your solution once" means that there will be only one message published in the topic `/red/plants_spots`, and "We run your solution N times" means that there will be `N` messages in that topic?

**Reply by allenthreee** [reply]
Sorry to bother you, but do you mean` /fruit_count` by `/red/plant_spots`?

**Reply by fpetric** [reply]
One run of your solution means we will execute start.sh script, wait until the UAV reports the fruit count or the time limit passes, kill the simulation. Each run starts with everything shut down (cold start as they call it). There will be only one message on the  `plant_beds` topic, and we expect only one message on the fruit count topic. 

To elaborate further, you will not receive another plant beds topic message once your UAV publishes the fruit count.

**Reply by allenthreee** [reply]
Hi Frano, 
The topic is still `/plants_beds` in the current GitHub code, and we are using `/plants_beds` all the time:), should we change it now? or just leave it?

**Reply by fpetric** [reply]
You can leave it, we can publish on both versions so it should be fine :)

**Reply by allenthreee** [reply]
Get it, thanks a lot:)

**Comment by pulak-gautam** [comment]
Hi, you mentioned you'd wait until the UAV reports the fruit count or the time limit passes, and kill the simulation. 
We have defined shell macros to wait until the plant_beds spawn in simulation, and then running actual submission (similar to what has been done in uav_ros_simulation package). 
Would the time taken for this spawning of plants be taken into account on the time limit as well, even though we haven't run any of the actual submission nodes until then?
Thank you

**Reply by fpetric** [reply]
Time starts when both challenge_started and plants_beds messages are published

**Comment by pulak-gautam** [comment]
Hi, just a small clarification, does the aforementioned time 11:59 AOE refer to 11:59am AOE or 11:59pm AOE?
Thank you

**Reply by fpetric** [reply]
It was intended as am, but pm will also be fine :)

**Comment by medini-sr** [comment]
When are we going to get feedback on our code?or are we going to get any?

**Reply by fpetric** [reply]
We are working on it. You will get feedback, yes
